app/uk/gov/hmrc/upscanuploadproxy/controllers/UploadController.scala

```diff
@@ -81,13 +81,24 @@ class UploadController @Inject()(uriGenerator: UploadUriGenerator, proxyService:
    * Otherwise send back the failure status code with a JSON body representing the error if possible.
    */
   private def handleFailure(errorAction: ErrorAction)(failure: FailureResponse): Result = {
-    val result = errorAction.redirectUrl.fold(
-      ifEmpty = Response.json(failure.statusCode, body = XmlErrorResponse.toJson(errorAction.key, failure.body))) {
-      redirectUrl => Response.redirect(redirectUrl, queryParams = XmlErrorResponse.toFields(errorAction.key, failure.body))
+    val xmlFields = XmlErrorResponse.toFields(errorAction.key, failure.body)
+    val result =  (errorAction.redirectUrl, isErrorRedirectUrlPolicyError(xmlFields, failure.statusCode)) match {
+      case (_, true) | (None, _)  => Response.json(failure.statusCode, body = XmlErrorResponse.toJson(errorAction.key, failure.body))
+      case (Some(errorRedirectUrl), _) => Response.redirect(errorRedirectUrl, queryParams = xmlFields)
     }
     result.withHeaders(failure.headers: _*)
   }
 
+  private def isErrorRedirectUrlPolicyError(xmlFields: Seq[(String, String)], failureStatusCode: Int): Boolean = {
+    lazy val isForbiddenStatusCode = failureStatusCode == FORBIDDEN
+    lazy val isAccessDeniedCode    = xmlFields.find(_._1 == "errorCode").exists(_._2 == "AccessDenied")
+    lazy val isErrorRedirectPolicyErrorMessage = xmlFields
+      .find(_._1 == "errorMessage")
+      .map(_._2)
+      .exists(m => m.contains("Policy Condition failed") && m.contains("$error_action_redirect"))
```

Comment:
is it not enough to just search for `error_action_redirect`? ie if the error is ever about `error_action_redirect` then assume we shouldn't redirect?

Reply:
Better to be sure of exact error. Also, I guess we don't expect this logic to be executed a lot. Most of the times we hope that we see success.